web/packages/teleport/src/components/EventRangePicker/Custom/Custom.tsx

```diff
@@ -0,0 +1,93 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useState, forwardRef } from 'react';
+import { isAfter, endOfDay, startOfDay, isSameDay } from 'date-fns';
+import { DayPicker, addToRange, DateRange } from 'react-day-picker';
+import 'react-day-picker/dist/style.css';
+
+import { StyledDateRange } from 'teleport/components/DayPicker/Shared';
+
+/**
+ * Allows user to select any range "from" (no limit)
+ * to "to" (no later than "today").
+ *
+ * @param currentRange is used to initially render the Calendar:
+ *   - if fields are undefined, the Calendar will render with
+ *     today's month with today's date bolded
+ *   - if fields are defined, the Calendar will render with the
+ *     provided dates with the range highlighted
+ */
+export const CustomRange = forwardRef<
```

Comment:
I think we could rename variables in this component to make it a bit easire to understand. At the moment it uses the `current` and `new` as prefixes in a way that goes against what I'm used to, but perhaps it's just me.

`currentRange` sounds like it's supposed to point at some kind of either constant or selected range, but it's used only for the initial selection.

`newRange` sounds like a variable that stores a new version of `range` that's soon going to be updated with `range`, but that's not what happens. `newRange` stores the current state of range within `CustomRange`.

`range` doesn't seem to carry any meaning since it's defined within `handleDayClick` and we know that it's `newRange` that actually holds the current range. So we have to read the code a little bit further to see that `range` is used as the new value for `newRange`.

With a couple of swaps, I think it becomes easier to understand:

* `currentRange` -> `initialRange`
* `newRange` -> `range`
* `range` -> `newRange`

Let me know what you think. If you want to apply the patch, copy it and then `pbpaste | git apply`.

<details>
<summary>Patch</summary>

```diff
diff --git a/web/packages/teleport/src/components/EventRangePicker/Custom/Custom.tsx b/web/packages/teleport/src/components/EventRangePicker/Custom/Custom.tsx
index 7adbd30db9..b0fc679889 100644
--- a/web/packages/teleport/src/components/EventRangePicker/Custom/Custom.tsx
+++ b/web/packages/teleport/src/components/EventRangePicker/Custom/Custom.tsx
@@ -27,7 +27,7 @@ import { StyledDateRange } from 'teleport/components/DayPicker/Shared';
  * Allows user to select any range "from" (no limit)
  * to "to" (no later than "today").
  *
- * @param currentRange is used to initially render the Calendar:
+ * @param initialRange is used to initially render the Calendar:
  *   - if fields are undefined, the Calendar will render with
  *     today's month with today's date bolded
  *   - if fields are defined, the Calendar will render with the
@@ -36,11 +36,11 @@ import { StyledDateRange } from 'teleport/components/DayPicker/Shared';
 export const CustomRange = forwardRef<
   HTMLDivElement,
   {
-    currentRange: DateRange;
+    initialRange: DateRange;
     onChange(from: Date, to: Date): void;
   }
->(({ currentRange, onChange }, ref) => {
-  const [newRange, setNewRange] = useState<DateRange | undefined>();
+>(({ initialRange, onChange }, ref) => {
+  const [range, setRange] = useState<DateRange | undefined>();
 
   function handleDayClick(selectedDay: Date) {
     // Don't let select date past today.
@@ -49,30 +49,30 @@ export const CustomRange = forwardRef<
     }
 
     // Don't do anything if `selected day` == `selected from`
-    if (newRange?.from && isSameDay(newRange.from, selectedDay)) {
+    if (range?.from && isSameDay(range.from, selectedDay)) {
       return;
     }
 
-    let range;
-    if (!newRange) {
+    let newRange;
+    if (!range) {
       // Start
-      range = addToRange(selectedDay, { from: undefined, to: undefined });
+      newRange = addToRange(selectedDay, { from: undefined, to: undefined });
     } else {
-      range = addToRange(selectedDay, { from: newRange.from, to: newRange.to });
+      newRange = addToRange(selectedDay, { from: range.from, to: range.to });
     }
 
-    if (range.from) {
-      range.from = startOfDay(range.from);
+    if (newRange.from) {
+      newRange.from = startOfDay(newRange.from);
     }
 
-    if (range.to) {
-      range.to = endOfDay(range.to);
+    if (newRange.to) {
+      newRange.to = endOfDay(newRange.to);
     }
 
-    setNewRange(range);
+    setRange(newRange);
 
-    if (range.from && range.to) {
-      onChange(range.from, range.to);
+    if (newRange.from && newRange.to) {
+      onChange(newRange.from, newRange.to);
     }
   }
 
@@ -81,11 +81,11 @@ export const CustomRange = forwardRef<
       <DayPicker
         mode="range"
         numberOfMonths={2}
-        defaultMonth={currentRange.from}
+        defaultMonth={initialRange.from}
         disabled={{
           after: new Date(),
         }}
-        selected={newRange || currentRange}
+        selected={range || initialRange}
         onDayClick={handleDayClick}
       />
     </StyledDateRange>
diff --git a/web/packages/teleport/src/components/EventRangePicker/EventRangePicker.tsx b/web/packages/teleport/src/components/EventRangePicker/EventRangePicker.tsx
index c0f82549ad..1351ebeade 100644
--- a/web/packages/teleport/src/components/EventRangePicker/EventRangePicker.tsx
+++ b/web/packages/teleport/src/components/EventRangePicker/EventRangePicker.tsx
@@ -77,7 +77,7 @@ export default function DataRange({ ml, range, onChangeRange, ranges }: Props) {
         open={isPickerOpen}
       >
         <CustomRange
-          currentRange={{ from: range.from, to: range.to }}
+          initialRange={{ from: range.from, to: range.to }}
           onChange={onSetCustomRange}
           ref={dayPickerRef}
         />

```
</details>

Reply:
We could also do

```typescript
const [newRange, setNewRange] = useState(currentRange);
```

which would then probably allow us to get rid of removing that `if (!newRange)` condition, but I haven't checked that.

Reply:
sorry isaiah, if you were testing the range picker from the `EventRangePicker.story.tsx`, that's a very broken story (i completely forgot to update, it's updated now)

Reply:
@kimlisa alright I took another stab, using the updated `EventRangePicker.story.tsx` (thanks!) to check that it appears to work properly:

<details>

```tsx
/*
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useState, forwardRef } from 'react';
import { isAfter, isBefore, endOfDay, startOfDay, isSameDay } from 'date-fns';
import { DayPicker, DateRange, DayClickEventHandler } from 'react-day-picker';
import 'react-day-picker/dist/style.css';

import { StyledDateRange } from 'teleport/components/DayPicker/Shared';

const isSameOrAfter = (dayClicked: Date, currentDay: Date): boolean => {
  return isSameDay(dayClicked, currentDay) || isAfter(dayClicked, currentDay);
};

const isSameOrBefore = (dayClicked: Date, currentDay: Date): boolean => {
  return isSameDay(dayClicked, currentDay) || isBefore(dayClicked, currentDay);
};

/**
 * Calculates the next date range based on the clicked day and the current range.
 *
 * @param {Date} dayClicked - The date that was clicked by the user.
 * @param {DateRange} currentRange - The current date range. It may have a defined 'from' date, 'to' date, both, or neither.
 *
 * @returns {DateRange} The new date range. The logic is as follows:
 *
 * If the current range has a defined 'from' but not a 'to':
 *   - If the clicked day is the same as or after the 'from' date, it becomes the 'to' date.
 *   - If the clicked day is before the 'from' date, it becomes the new 'from' date and the old 'from' date becomes the 'to' date.
 *
 * If the current range has a defined 'to' but not a 'from':
 *   - If the clicked day is the same as or before the 'to' date, it becomes the 'from' date.
 *   - If the clicked day is after the 'to' date, it becomes the new 'to' date and the old 'to' date becomes the 'from' date.
 *
 * If the current range is either undefined or fully defined, a new range is started with the clicked day as the 'from' date and 'to' date as undefined.
 */
const calcNextRange = (
  dayClicked: Date,
  currentRange?: DateRange
): DateRange => {
  // If the current range has a defined `from` but not a defined `to`...
  if (currentRange.from && !currentRange.to) {
    // If the `dayClicked` >= `currentRange.from`, make it the `to`.
    if (isSameOrAfter(dayClicked, currentRange.from)) {
      return { from: currentRange.from, to: endOfDay(dayClicked) };
    }
    // If the `dayClicked` < `currentRange.from`, make it the `from`
    // and the `currentRange.from` the `to`.
    return { from: startOfDay(dayClicked), to: endOfDay(currentRange.from) };
  }

  // If the current range has a defined `to` but not a defined `from`...
  if (currentRange.to && !currentRange.from) {
    // If the `dayClicked` <= `currentRange.to`, make it the `from`.
    if (isSameOrBefore(dayClicked, currentRange.to)) {
      return { from: startOfDay(dayClicked), to: currentRange.to };
    }
    // If the `dayClicked` > `currentRange.to`, make it the `to`
    // and the `currentRange.to` the `from`.
    return { from: startOfDay(currentRange.to), to: endOfDay(dayClicked) };
  }

  // Otherwise the current range is undefined or fully defined,
  // in which case start a new range `from` the `dayClicked`.
  return { from: startOfDay(dayClicked), to: undefined };
};

/**
 * Allows user to select any range "from" (no limit)
 * to "to" (no later than "today").
 *
 * @param initialRange can be used to set the initial range.
 * @param onFullyDefined is called when the user has selected a "from" and "to" date.
 */
export const CustomRange = forwardRef<
  HTMLDivElement,
  {
    initialRange?: DateRange;
    onFullyDefined(from: Date, to: Date): void;
  }
>(({ initialRange, onFullyDefined }, ref) => {
  const [currentRange, setCurrentRange] = useState<DateRange | undefined>(
    initialRange
  );

  const handleDayClick: DayClickEventHandler = dayClicked => {
    let newRange = calcNextRange(dayClicked, currentRange);

    // Set the component's state to the newly calculated range.
    setCurrentRange(newRange);
    // If the newly calculated range is fully defined, call the callback.
    if (newRange.from && newRange.to) {
      onFullyDefined(newRange.from, newRange.to);
    }
  };

  return (
    <StyledDateRange ref={ref}>
      <DayPicker
        mode="range"
        numberOfMonths={2}
        defaultMonth={initialRange.from}
        disabled={{
          after: new Date(),
        }}
        selected={currentRange}
        onDayClick={handleDayClick}
      />
    </StyledDateRange>
  );
});
```

</details>

This is easier to follow imo, I was getting lost in the previous one, but of course it's up to you whether you agree.

Reply:
i kept it as is 😅 (or rather i shorten it even more: https://github.com/gravitational/teleport/pull/36319/commits/9c25717bf77f4bd80bb9d85df8e3837f589daac6)